README.md
[<img src="https://api.travis-ci.org/mnmelo/lazy_import.svg">](https://travis-ci.org/mnmelo/lazy_import)

# lazy_import
A module for lazy loading of Python modules


CHANGELOG
??/??/??      v0.2.3
 - Ensure that when a package of a lazy-loaded module is loaded, the
   lazy-loaded module remains an attribute of the newly loaded package.
 - Ensure that when two lazy-loaded modules exist in the same lazy package,
   the second (and all subsequent) lazy-modules are also added to the lazy
   packages list of lazy sub-modules.
 - More informative error when inadvertently trying to use a lazy callable a
   baseclass (Issue #2).
 - Extra level of debugging (trace debugging), toggled via lazy_import.logger.
22/01/2018    v0.2.2
 - fixed a serious bug when lazy-loading a submodule of a fully-loaded base.
17/01/2018    v0.2.1
 - annoying release to fix install notice on PyPi (because README.rst was
   incorrect).
17/01/2018    v0.2
 - stabilized a python 3 bug when importing modules using 'import modulename'
   syntax.
 - added a test shortcut.
 - published to PyPi (yay!)

25/08/2017    v0.1 
 - along with python 2.7, the code is now Python 3 compatible
   (at least versions 3.4 through 3.6).
 - a replacement of lazyModule (lazy_module, which defers most work to
   _lazy_module) was implemented that uses a modified LazyModule class;
 - a different LazyModule class is now created per instance, so that
   reverting the __getattribute__ behavior can be done safely;
 - a function to lazily import module callables was added;
 - customization is fully supported, either as passable custom error messages
   or as passable LazyModule subclass alternatives.
 - module reloading is now more stable, especially in the case of submodules.
 - added testing.



build_and_test.sh
#!/bin/bash -e
basedir=$(pwd)
cd
pip3 install --force-reinstall --user $basedir
pip2 install --force-reinstall --user $basedir
python3 -m pytest --boxed -n 4 -v --pyargs lazy_import
python2 -m pytest --boxed -n 4 -v --pyargs lazy_import
cd $basedir


lazy_import/test_lazy.py
# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
#
# Testing for lazy_import --- https://github.com/mnmelo/lazy_import
# Copyright (C) 2017-2018 Manuel Nuno Melo
#
# This file is part of lazy_import.
#
#  lazy_import is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  lazy_import is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with lazy_import.  If not, see <http://www.gnu.org/licenses/>.
#

import pytest
import importlib
import sys
import six
from six.moves import range
import itertools
import string
import random
random.seed(42) # For consistency when parallel-testing

import lazy_import

# Shortcut to allow direct testing from within python

def run(np=None):
    if np is None:
        import multiprocessing
        np = multiprocessing.cpu_count()
    import os
    path = os.path.dirname(__file__)
    if np > 1:
        pytest.main(['-n', str(np), '--boxed', path])
    else:
        pytest.main(['--boxed', path])
###############################################################################
# Constants and util functions

RANDOM_PREFIX = "_lazy_test_"
RANDOM_ALPHABET = string.ascii_letters + string.digits
def random_str(size=5):
    # We make it start with a letter
    return "m" + ''.join(random.choice(RANDOM_ALPHABET) for _ in range(size))

def random_modname(submodules=0):
    while True:
        subname = ".".join(random_str() for _ in range(submodules + 1))
        modname = RANDOM_PREFIX + subname
        if modname in _GENERATED_NAMES or modname in sys.modules:
            continue
        _GENERATED_NAMES.append(modname)
        return modname 

class _TestLazyModule(lazy_import.LazyModule):
    pass

_GENERATED_NAMES = []
# Modules not usually loaded on startup. Must include at least one with
#  submodule
NAMES_EXISTING = ("sched", "distutils.core")

LEVELS = ("leaf", "base")
CLASSES = (_TestLazyModule, lazy_import.LazyModule)
CALLABLE_NAMES =("fn1", ("fn1"), ("fn1", "fn2"), ("fn1", "fn2", "fn3"))
ERROR_MSGS = (None, {"msg":"Module: {module}\n"
                           "Caller: {caller}\n"
                           "Install Name: {install_name}\n",
                     "msg_callable":"Module: {module}\n"
                           "Caller: {caller}\n"
                           "Install Name: {install_name}\n"
                           "Callable: {callable}\n",
                     "module":"error_modname",
                     "caller":"error_callername",
                     "install_name":"error_installname"})
CALLABLE_ALIASES = (lazy_import.lazy_callable,
        lazy_import.lazy_function, lazy_import.lazy_class)

###############################################################################
# Fixtures

@pytest.fixture(params=itertools.product(LEVELS, CLASSES, ERROR_MSGS))
def lazy_opts(request):
    return request.param

###############################################################################
# Re-usable tests

def _check_reuse(modname):
    curr_module_id = id(sys.modules[modname])
    newmod = importlib.import_module(modname)
    assert id(newmod) == curr_module_id

def _check_lazy_loading(modname):
    names = modname.split(".")
    _check_not_loaded(modname)
    basename = lazy_import.module_basename(modname)
    mod = lazy_import.lazy_module(modname, level="leaf")
    assert sys.modules[modname] is mod
    assert str(mod) == "Lazily-loaded module " + modname
    base = lazy_import.lazy_module(modname, level="base")
    assert sys.modules[basename] is base
    assert str(base) == "Lazily-loaded module " + basename

    # A subtle bug only rears its head if an actual import statement is done.
    # Can only test that for the module names we know of
    # (meaning, no random ones)
    if modname in NAMES_EXISTING:
        if modname == 'sched':
            import sched as newmod
        elif modname == 'distutils.core':
            import distutils.core as newmod
        assert str(newmod) == "Lazily-loaded module " + modname

    # Check that all submodules are in and that submodule access works
    curr_name = basename
    curr_mod = base
    for submod in names[1:]:
        curr_name += "." + submod
        curr_mod = getattr(curr_mod, submod)
        assert curr_name in sys.modules
        assert str(curr_mod) == "Lazily-loaded module " + curr_name
        assert isinstance(curr_mod, lazy_import.LazyModule)
        _check_reuse(curr_name)
        # Check that missing modules raise errors
        if modname in _GENERATED_NAMES:
            _check_module_missing(curr_mod)

def _check_module_missing(obj, msg=None, call=False):
    with pytest.raises(ImportError) as excinfo:
        if call:
            obj()
        else:
            obj.modattr
    if msg is not None:
        assert str(excinfo.value) == msg

def _check_callable_missing(obj, msg=None):
    with pytest.raises(AttributeError) as excinfo:
        obj()
    if msg is not None:
        assert str(excinfo.value) == msg

def _check_not_loaded(modname):
    assert modname not in sys.modules, \
        modname + " already loaded. Maybe use with pytest's xdist's '--boxed'?"

###############################################################################
# TESTS                              TESTS                              TESTS #
###############################################################################

@pytest.mark.parametrize("modname", tuple(random_modname(i) for i in range(3))
                                    + NAMES_EXISTING)
def test_lazyload(modname):
    _check_lazy_loading(modname)

def test_presentload():
    import os.path
    mod = lazy_import.lazy_module("os")
    assert mod is os
    mod = lazy_import.lazy_module("os.path")
    assert mod is os.path
    mod = lazy_import.lazy_module("os.path", level="base")
    assert mod is os
    assert not isinstance(mod, lazy_import.LazyModule)

@pytest.mark.parametrize("nsub", range(3))
def test_opts(nsub, lazy_opts):
    modname = random_modname(nsub)
    level, modclass, errors = lazy_opts
    mod = lazy_import.lazy_module(modname, error_strings=errors,
            lazy_mod_class=modclass, level=level)
    names = modname.split(".")
    basename = lazy_import.module_basename(modname)
    if level == "leaf":
        assert sys.modules[modname] is mod
        err_modname = modname
    elif level == "base":
        assert sys.modules[basename] is mod
        err_modname = basename
    else:
        raise ValueError("Unexpected value {} for 'level'".format(level))
    # Test the exception err msg
    assert isinstance(mod, modclass)
    if errors is None:
        expected_err = lazy_import._MSG.format(module=err_modname,
                                               caller=__name__,
                                               install_name=basename)
    else:
        expected_err = errors["msg"].format(**errors)
    _check_module_missing(mod, msg=expected_err)


@pytest.mark.parametrize("nsub, errors, cnames, fn",
        itertools.product(range(3), ERROR_MSGS, CALLABLE_NAMES,
                          CALLABLE_ALIASES))
def test_callable_missing_module(nsub, errors, cnames, fn):
    modname = random_modname(nsub)
    basename = lazy_import.module_basename(modname)
    if isinstance(cnames, six.string_types):
        lazys = (fn(modname+"."+cnames, error_strings=errors),)
        cnames = (cnames, )
    else:
        lazys = fn(modname, *cnames, error_strings=errors)
    for lazy in lazys:
        if errors is None:
            expected_err = lazy_import._MSG.format(module=modname,
                                                   caller=__name__,
                                                   install_name=basename)
        else:
            expected_err = errors["msg"].format(**errors)
        _check_module_missing(lazy, msg=expected_err, call=True)
    
@pytest.mark.parametrize("modname, errors, cnames, fn",
        itertools.product(NAMES_EXISTING, ERROR_MSGS, CALLABLE_NAMES,
                          CALLABLE_ALIASES))
def test_callable_missing(modname, errors, cnames, fn):
    _check_not_loaded(modname)
    basename = lazy_import.module_basename(modname)
    if isinstance(cnames, six.string_types):
        lazys = (fn(modname+"."+cnames, error_strings=errors),)
        cnames = (cnames, )
    else:
        lazys = fn(modname, *cnames, error_strings=errors)
    for lazy, cname in zip(lazys, cnames):
        if errors is None:
            expected_err = lazy_import._MSG_CALLABLE.format(module=modname,
                                                   caller=__name__,
                                                   install_name=basename,
                                                   callable=cname)
        else:
            errors['callable'] = cname
            expected_err = errors["msg_callable"].format(**errors)
        _check_callable_missing(lazy, msg=expected_err)
    
@pytest.mark.parametrize("modname, errors, cnames, fn",
        itertools.product(NAMES_EXISTING, ERROR_MSGS, CALLABLE_NAMES,
                          CALLABLE_ALIASES))
def test_error_callable_as_baseclass(modname, errors, cnames, fn):
    _check_not_loaded(modname)
    if isinstance(cnames, six.string_types):
        lazys = (fn(modname+"."+cnames, error_strings=errors),)
    else:
        lazys = fn(modname, *cnames, error_strings=errors)
    for lazy in lazys:
        with pytest.raises(NotImplementedError) as excinfo:
            class TestClass(lazy):
                pass

@pytest.mark.parametrize("modname",
                [modname for modname in NAMES_EXISTING if '.' in modname])
def test_load_lazysub_on_fullmodule(modname, lazy_opts):
    level, modclass, errors = lazy_opts
    base = modname.split('.')[0]
    # A fully loaded base...
    importlib.import_module(base)
    # and a lazy sub...
    mod = lazy_import.lazy_module(modname, error_strings=errors,
                                   lazy_mod_class=modclass, level=level)
    if level == 'base':
        assert not isinstance(mod, modclass)
    else:
        assert isinstance(mod, modclass)


lazy_import/lazy_import.py
# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
#
# lazy_import --- https://github.com/mnmelo/lazy_import
# Copyright (c) 2017 Manuel Nuno Melo
#
# Released under the GNU Public Licence, v3 or any higher version.
#
# This module was based on code from the importing module from the PEAK
# package (see http://peak.telecommunity.com/DevCenter/FrontPage). The PEAK
# package is released under the following license, reproduced here:
#
#  Copyright (C) 1996-2004 by Phillip J. Eby and Tyler C. Sarna.
#  All rights reserved.  This software may be used under the same terms
#  as Zope or Python.  THERE ARE ABSOLUTELY NO WARRANTIES OF ANY KIND.
#  Code quality varies between modules, from "beta" to "experimental
#  pre-alpha".  :)
#
# The following list summarizes the modifications to the PEAK importing code:
#  - the code is now Python 3 compatible (at least versions 3.4 through 3.6).
#  - a replacement of lazyModule (lazy_module, which defers most work to
#    _lazy_module) was implemented that uses a modified LazyModule class;
#  - a different LazyModule class is now created per instance, so that
#    reverting the __getattribute__ behavior can be done safely;
#  - a function to lazily import module callables was added;
#  - customization is fully supported, either as passable custom error messages
#    or as passable LazyModule subclass alternatives.

"""
Lazy module loading
===================

Functions and classes for lazy module loading that also delay import errors.
Heavily borrowed from the `importing`_ module.

.. _`importing`: http://peak.telecommunity.com/DevCenter/Importing

Files and directories
---------------------

.. autofunction:: module
.. autofunction:: callable

"""

__all__ = ['lazy_module', 'lazy_callable', 'lazy_function', 'lazy_class',
           'LazyModule', 'LazyCallable', 'module_basename', '_MSG',
           '_MSG_CALLABLE']

from types import ModuleType
import sys
try:
    # imp is deprecated since python 3.4 but there's no clear alternative to
    # the lock mechanism, other than to import directly from _imp.
    from imp import acquire_lock, release_lock 
except ImportError:
    from _imp import acquire_lock, release_lock 

# Adding a __spec__ doesn't really help. I'll leave the code here in case
# future python implementations start relying on it.
#try:
#    from importlib.machinery import ModuleSpec
#except ImportError:
#    ModuleSpec = None

import six
from six import raise_from
from six.moves import reload_module

_DEBUG = False
if _DEBUG:
    import traceback

################################
# Module/function registration #
################################

#### Lazy classes ####

class LazyModule(ModuleType):
    """Class for lazily-loaded modules that triggers proper loading on access.

    Instantiation should be made from a subclass of :class:`LazyModule`, with
    one subclass per instantiated module. Regular attribute set/access can then
    be recovered by setting the subclass's :meth:`__getattribute__` and
    :meth:`__setattribute__` to those of :class:`types.ModuleType`.
    """
    # peak.util.imports sets __slots__ to (), but it seems pointless because
    # the base ModuleType doesn't itself set __slots__.
    def __getattribute__(self, attr):
        if _DEBUG:
            sys.stderr.write("Getting attr {} of LazyModule instance of {}\n"
                    .format(attr, super(LazyModule, self).
                        __getattribute__("__name__")))
            #traceback.print_stack()
        # IPython tries to be too clever and constantly inspects, asking for
        #  modules' attrs, which causes premature module loading and unesthetic
        #  internal errors if the lazily-loaded module doesn't exist.
        if (run_from_ipython()
            and (attr.startswith(("__", "_ipython"))
                 or attr == "_repr_mimebundle_")
            and module_basename(_caller_name()) in ('inspect', 'IPython')):
                raise AttributeError
        if not attr in ('__name__','__class__'):
            # If it's an already-loaded submodule, we return it without
            # triggering a full loading
            try:
                return sys.modules[self.__name__+"."+attr]
            except KeyError:
                pass
            # Check if it's one of the lazy callables
            try:
                return type(self)._lazy_import_callables[attr]
            except (AttributeError, KeyError):
                _load_module(self)
        return super(LazyModule, self).__getattribute__(attr)

    def __setattr__(self, attr, value):
        if _DEBUG:
            sys.stderr.write("Setting attr {} to value {}, "
                             "in LazyModule instance of {}\n"
                             .format(attr, value, super(LazyModule, self).
                             __getattribute__("__name__")))
        _load_module(self)
        return super(LazyModule, self).__setattr__(attr, value)


class LazyCallable(object):
    """Class for lazily-loaded callables that triggers module loading on access

    """
    def __init__(self, module, cname):
        self.module = module
        self.modclass = type(self.module)
        self.cname = cname
        self.callable = None
        # Need to save these, since the module-loading gets rid of them
        self.error_msgs = self.modclass._lazy_import_error_msgs
        self.error_strings = self.modclass._lazy_import_error_strings

    def __call__(self, *args, **kwargs):
        # No need to go through all the reloading more than once.
        if self.callable:
            return self.callable(*args, **kwargs)
        try:
            del self.modclass._lazy_import_callables[self.cname]
        except (AttributeError, KeyError):
            pass
        try:
            self.callable = getattr(self.module, self.cname)
        except AttributeError:
            msg = self.error_msgs['msg_callable']
            raise_from(AttributeError(
                msg.format(callable=self.cname, **self.error_strings)), None)
        except ImportError as err:
            # Import failed. We reset the dict and re-raise the ImportError.
            try:
                self.modclass._lazy_import_callables[self.cname] = self
            except AttributeError:
                self.modclass._lazy_import_callables = {self.cname: self}
            raise_from(err, None)
        else:
            return self.callable(*args, **kwargs)


### Functions ###

def lazy_module(modname, error_strings=None, lazy_mod_class=LazyModule,
                  level='leaf'):
    """Function allowing lazy importing of a module into the namespace.

    A lazy module object is created, registered in `sys.modules`, and
    returned. This is a hollow module; actual loading, and `ImportErrors` if
    not found, are delayed until an attempt is made to access attributes of the
    lazy module.

    A handy application is to use :func:`lazy_module` early in your own code
    (say, in `__init__.py`) to register all modulenames you want to be lazy.
    Because of registration in `sys.modules` later invocations of
    `import modulename` will also return the lazy object. This means that after
    initial registration the rest of your code can use regular pyhon import
    statements and retain the lazyness of the modules.

    Parameters
    ----------
    modname : str
         The module to import.
    error_strings : dict, optional
         A dictionary of strings to use when module-loading fails. Key 'msg'
         sets the message to use (defaults to :attr:`lazy_import._MSG`). The
         message is formatted using the remaining dictionary keys. The default
         message informs the user of which module is missing (key 'module'),
         what code loaded the module as lazy (key 'caller'), and which package
         should be installed to solve the dependency (key 'install_name').
         None of the keys is mandatory and all are given smart names by default.
    lazy_mod_class: type, optional
         Which class to use when instantiating the lazy module, to allow
         deep customization. The default is :class:`LazyModule` and custom
         alternatives **must** be a subclass thereof.
    level : str, optional
         Which submodule reference to return. Either a reference to the 'leaf'
         module (the default) or to the 'base' module. This is useful if you'll
         be using the module functionality in the same place you're calling
         :func:`lazy_module` from, since then you don't need to run `import`
         again. Setting *level* does not affect which names/modules get
         registered in `sys.modules`.
         For *level* set to 'base' and *modulename* 'aaa.bbb.ccc'::

            aaa = lazy_import.lazy_module("aaa.bbb.ccc", level='base')
            # 'aaa' becomes defined in the current namespace, with
            #  (sub)attributes 'aaa.bbb' and 'aaa.bbb.ccc'.
            # It's the lazy equivalent to:
            import aaa.bbb.ccc

        For *level* set to 'leaf'::

            ccc = lazy_import.lazy_module("aaa.bbb.ccc", level='leaf')
            # Only 'ccc' becomes set in the current namespace.
            # Lazy equivalent to:
            from aaa.bbb import ccc

    Returns
    -------
    module
        The module specified by *modname*, or its base, depending on *level*.
        The module isn't immediately imported. Instead, an instance of
        *lazy_mod_class* is returned. Upon access to any of its attributes, the
        module is finally loaded.

    Examples
    --------
    >>> import lazy_import, sys
    >>> np = lazy_import.lazy_module("numpy")
    >>> np
    Lazily-loaded module numpy
    >>> np is sys.modules['numpy']
    True
    >>> np.pi # This causes the full loading of the module ...
    3.141592653589793
    >>> np # ... and the module is changed in place. 
    <module 'numpy' from '/usr/local/lib/python/site-packages/numpy/__init__.py'>

    >>> import lazy_import, sys
    >>> # The following succeeds even when asking for a module that's not available
    >>> missing = lazy_import.lazy_module("missing_module")
    >>> missing
    Lazily-loaded module missing_module
    >>> missing is sys.modules['missing_module']
    True
    >>> missing.some_attr # This causes the full loading of the module, which now fails.
    ImportError: __main__ attempted to use a functionality that requires module missing_module, but it couldn't be loaded. Please install missing_module and retry.

    See Also
    --------
    :func:`lazy_callable`
    :class:`LazyModule`

    """
    if error_strings is None:
        error_strings = {}
    _set_default_errornames(modname, error_strings)

    mod = _lazy_module(modname, error_strings, lazy_mod_class)
    if level == 'base':
        return sys.modules[module_basename(modname)]
    elif level == 'leaf':
        return mod
    else:
        raise ValueError("Parameter 'level' must be one of ('base', 'leaf')")


def _lazy_module(modname, error_strings, lazy_mod_class):
    acquire_lock()
    try:
        fullmodname = modname
        fullsubmodname = None
        # ensure parent module/package is in sys.modules
        # and parent.modname=module, as soon as the parent is imported   
        while modname:
            try:
                mod = sys.modules[modname]
                # We reached a (base) module that's already loaded. Let's stop
                # the cycle.
                modname = ''
            except KeyError:
                err_s = error_strings.copy()
                err_s.setdefault('module', modname)

                class _LazyModule(lazy_mod_class):
                    _lazy_import_error_msgs = {'msg': err_s.pop('msg')}
                    try:
                        _lazy_import_error_msgs['msg_callable'] = \
                                err_s.pop('msg_callable')
                    except KeyError:
                        pass
                    _lazy_import_error_strings = err_s
                    _lazy_import_callables = {}

                    def __repr__(self):
                        return "Lazily-loaded module {}".format(self.__name__)
                # A bit of cosmetic, to make AttributeErrors read more natural  
                _LazyModule.__name__ = 'module'
                # Actual module instantiation
                mod = sys.modules[modname] = _LazyModule(modname)
                # No need for __spec__. Maybe in the future.
                #if ModuleSpec:
                #    ModuleType.__setattr__(mod, '__spec__',
                #            ModuleSpec(modname, None))
            if fullsubmodname:
                ModuleType.__setattr__(mod, submodname,
                                       sys.modules[fullsubmodname])
            fullsubmodname = modname
            modname, _, submodname = modname.rpartition('.')
        return sys.modules[fullmodname]
    finally:
        release_lock()


def lazy_callable(modname, *names, **kwargs):
    """Performs lazy importing of one or more callables.

    :func:`lazy_callable` creates functions that are thin wrappers that pass
    any and all arguments straight to the target module's callables. These can
    be functions or classes. The full loading of that module is only actually
    triggered when the returned lazy function itself is called. This lazy
    import of the target module uses the same mechanism as
    :func:`lazy_module`.
    
    If, however, the target module has already been fully imported prior
    to invocation of :func:`lazy_callable`, then the target callables
    themselves are returned and no lazy imports are made.

    :func:`lazy_function` and :func:`lazy_function` are aliases of
    :func:`lazy_callable`.

    Parameters
    ----------
    modname : str
         The base module from where to import the callable(s) in *names*,
         or a full 'module_name.callable_name' string.
    names : str (optional)
         The callable name(s) to import from the module specified by *modname*.
         If left empty, *modname* is assumed to also include the callable name
         to import.
    error_strings : dict, optional
         A dictionary of strings to use when reporting loading errors (either a
         missing module, or a missing callable name in the loaded module).
         *error_string* follows the same usage as described under
         :func:`lazy_module`, with the exceptions that 1) a further key,
         'msg_callable', can be supplied to be used as the error when a module
         is successfully loaded but the target callable can't be found therein
         (defaulting to :attr:`lazy_import._MSG_CALLABLE`); 2) a key 'callable'
         is always added with the callable name being loaded.
    lazy_mod_class : type, optional
         See definition under :func:`lazy_module`.
    lazy_call_class : type, optional
         Analogously to *lazy_mod_class*, allows setting a custom class to
         handle lazy callables, other than the default :class:`LazyCallable`.

    Returns
    -------
    wrapper function or tuple of wrapper functions
        If *names* is passed, returns a tuple of wrapper functions, one for
        each element in *names*.
        If only *modname* is passed it is assumed to be a full
        'module_name.callable_name' string, in which case the wrapper for the
        imported callable is returned directly, and not in a tuple.
        
    Notes
    -----
    Unlike :func:`lazy_module`, which returns a lazy module that eventually
    mutates into the fully-functional version, :func:`lazy_callable` only
    returns thin wrappers that never change. This means that the returned
    wrapper object never truly becomes the one under the module's namespace,
    even after successful loading of the module in *modname*. This is fine for
    most practical use cases, but may break code that relies on the usage of
    the returned objects oter than calling them. One such example is the lazy
    import of a class: it's fine to use the returned wrapper to instantiate an
    object, but it can't be used, for instance, to subclass from.

    Examples
    --------
    >>> import lazy_import, sys
    >>> fn = lazy_import.lazy_callable("numpy.arange")
    >>> sys.modules['numpy']
    Lazily-loaded module numpy
    >>> fn(10)
    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
    >>> sys.modules['numpy']
    <module 'numpy' from '/usr/local/lib/python3.5/site-packages/numpy/__init__.py'>

    >>> import lazy_import, sys
    >>> cls = lazy_import.lazy_callable("numpy.ndarray") # a class
    >>> class MyArray(cls): # FAIL!
    >>>     pass

    See Also
    --------
    :func:`lazy_module`
    :class:`LazyCallable`
    :class:`LazyModule`

    """
    if not names:
        modname, _, name = modname.rpartition(".")
    lazy_mod_class = _setdef(kwargs, 'lazy_mod_class', LazyModule)
    lazy_call_class = _setdef(kwargs, 'lazy_call_class', LazyCallable)
    error_strings = _setdef(kwargs, 'error_strings', {})
    _set_default_errornames(modname, error_strings, call=True)

    if not names:
        # We allow passing a single string as 'modname.callable_name',
        # in which case the wrapper is returned directly and not as a list.
        return _lazy_callable(modname, name, error_strings.copy(),
                                lazy_mod_class, lazy_call_class)
    return tuple(_lazy_callable(modname, cname, error_strings.copy(),
                        lazy_mod_class, lazy_call_class) for cname in names)

lazy_function = lazy_class = lazy_callable

def _lazy_callable(modname, cname, error_strings,
                     lazy_mod_class, lazy_call_class):
    # We could do most of this in the LazyCallable __init__, but here we can
    # pre-check whether to actually be lazy or not.
    module = _lazy_module(modname, error_strings, lazy_mod_class)
    modclass = type(module)
    if (issubclass(modclass, LazyModule) and
        hasattr(modclass, '_lazy_import_callables')):
        modclass._lazy_import_callables.setdefault(
            cname, lazy_call_class(module, cname))
    return getattr(module, cname)


#######################
# Real module loading #
#######################

def _load_module(module):
    """Ensures that a module, and its parents, are properly loaded

    """
    modclass = type(module)
    # We only take care of our own LazyModule instances
    if not issubclass(modclass, LazyModule):
        return
    acquire_lock()
    try:
        # We first identify whether this is a loadable LazyModule, then we
        # strip as much of lazy_import behavior as possible (keeping it cached,
        # in case loading fails and we need to reset the lazy state).
        if not hasattr(modclass, '_lazy_import_error_msgs'):
            # Alreay loaded (no _lazy_import_error_msgs attr). Not reloading.
            return
        cached_data = _clean_lazymodule(modclass)

        # First, ensure the parent is loaded (using recursion; *very* unlikely
        # we'll ever hit a stack limit in this case).
        parent, _, modname = module.__name__.rpartition('.')
        try:
            if parent:
                _load_module(sys.modules[parent])
                setattr(sys.modules[parent], modname, module)
            # Get Python to do the real import!
            reload_module(module)           
        except:
            # Loading failed. We reset our lazy state.
            if _DEBUG:
                sys.stderr.write("**InExcept!\n")
            _reset_lazymodule(modclass, cached_data)
            raise
    except (AttributeError, ImportError) as err:
        if _DEBUG:
            sys.stderr.write("**InOuterExcept!\n")
            print(err)
            #traceback.print_stack()
        # Under Python 3 reloading our dummy LazyModule instances causes an
        # AttributeError if the module can't be found. Would be preferrable if
        # we could always rely on an ImportError. As it is we vet the
        # AttributeError as thoroughly as possible.
        if ((six.PY3 and isinstance(err, AttributeError)) and not
                 err.args[0] == "'NoneType' object has no attribute 'name'"):
            # Not the AttributeError we were looking for.
            raise
        msg = modclass._lazy_import_error_msgs['msg']
        raise_from(ImportError(
            msg.format(**modclass._lazy_import_error_strings)), None)
    finally:
        release_lock()


##############################
# Helper functions/constants #
##############################

_MSG = ("{caller} attempted to use a functionality that requires module "
        "{module}, but it couldn't be loaded. Please install {install_name} "
        "and retry.")

_MSG_CALLABLE = ("{caller} attempted to use a functionality that requires "
           "{callable}, of module {module}, but it couldn't be found in that "
           "module. Please install a version of {install_name} that has "
           "{module}.{callable} and retry.")

_CLS_ATTRS = ("_lazy_import_error_strings", "_lazy_import_error_msgs",
              "_lazy_import_callables", "__repr__")

def _setdef(argdict, name, defaultvalue):
    """Like dict.setdefault but sets the default value also if None is present.

    """
    if not name in argdict or argdict[name] is None:
        argdict[name] = defaultvalue
    return argdict[name]


def module_basename(modname):
    return modname.partition('.')[0]


def _set_default_errornames(modname, error_strings, call=False):
    # We don't set the modulename default here because it will change for
    # parents of lazily imported submodules.
    error_strings.setdefault('caller', _caller_name(3, default='Python'))
    error_strings.setdefault('install_name', module_basename(modname))
    error_strings.setdefault('msg', _MSG)
    if call:
        error_strings.setdefault('msg_callable', _MSG_CALLABLE)


def _caller_name(depth=2, default=''):
    """Returns the name of the calling namespace.

    """
    # the presence of sys._getframe might be implementation-dependent.
    # It isn't that serious if we can't get the caller's name.
    try:
        return sys._getframe(depth).f_globals['__name__']
    except AttributeError:
        return default


def _clean_lazymodule(modclass):
    """Removes all lazy behavior from a module's class, for loading.

    Returns
    -------
    dict
        A dictionary of deleted class attributes, that can be used to reset the
        lazy state using :func:`_reset_lazymodule`.
    """
    modclass.__getattribute__ = ModuleType.__getattribute__
    modclass.__setattr__ = ModuleType.__setattr__
    cls_attrs = {}
    for cls_attr in _CLS_ATTRS:
        try:
            cls_attrs[cls_attr] = getattr(modclass, cls_attr)
            delattr(modclass, cls_attr)
        except AttributeError:
            pass
    return cls_attrs


def _reset_lazymodule(modclass, cls_attrs):
    """Resets a module's lazy state from cached data.

    """
    del modclass.__getattribute__
    del modclass.__setattr__
    for cls_attr in _CLS_ATTRS:
        try:
            setattr(modclass, cls_attr, cls_attrs[cls_attr])
        except KeyError:
            pass


def run_from_ipython():
    # Taken from https://stackoverflow.com/questions/5376837
    try:
        __IPYTHON__
        return True
    except NameError:
        return False


